### ISSUE TITLE
🔥 Not getting IOS notifications when background or killed

### ISSUE THREAD

**Issue by harrydema** [invertase/react-native-firebase#3969]
## Issue

I am gettting crazy to make it work on background or quit mode on IOS. The notifications do work on foreground.

The notifications arrive as soon as I get into the app.

### Javascript

<details><summary>Click To Expand</summary>
<p>

#### `package.json`:

```json
{
  "name": "TestPN",
  "version": "0.0.1",
  "private": true,
  "scripts": {
    "android": "react-native run-android",
    "ios": "react-native run-ios",
    "start": "react-native start",
    "test": "jest",
    "lint": "eslint ."
  },
  "dependencies": {
    "@notifee/react-native": "^0.9.1",
    "@react-native-community/clipboard": "^1.2.3",
    "@react-native-firebase/app": "^8.2.0",
    "@react-native-firebase/messaging": "^7.4.2",
    "react": "16.13.1",
    "react-native": "0.63.1"
  },
  "devDependencies": {
    "@babel/core": "^7.8.4",
    "@babel/runtime": "^7.8.4",
    "@react-native-community/eslint-config": "^1.1.0",
    "babel-jest": "^25.1.0",
    "eslint": "^6.5.1",
    "jest": "^25.1.0",
    "metro-react-native-babel-preset": "^0.59.0",
    "react-test-renderer": "16.13.1"
  },
  "jest": {
    "preset": "react-native"
  }
}
```

#### `firebase.json` for react-native-firebase v6:

```json
# N/A
```

#### `App.js`
```js
/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 *
 * @format
 * @flow strict-local
 */

import React, { useEffect } from 'react';
import messaging from '@react-native-firebase/messaging';
import {
  SafeAreaView,
  StyleSheet,
  ScrollView,
  View,
  Text,
  StatusBar,
} from 'react-native';
import Clipboard from '@react-native-community/clipboard'

import {
  Header,
  LearnMoreLinks,
  Colors,
  DebugInstructions,
  ReloadInstructions,
} from 'react-native/Libraries/NewAppScreen';

const App: () => React$Node = () => {
  useEffect(() => {
    // Get the device token
    messaging()
      .getToken()
      .then(token => {
        return Clipboard.setString(token);
      });

    // Listen to whether the token changes
    return messaging().onTokenRefresh(token => {
      return Clipboard.setString(token);
    });
  }, []);
  return (
    <>
      <StatusBar barStyle="dark-content" />
      <SafeAreaView>
        <ScrollView
          contentInsetAdjustmentBehavior="automatic"
          style={styles.scrollView}>
          <Header />
          {global.HermesInternal == null ? null : (
            <View style={styles.engine}>
              <Text style={styles.footer}>Engine: Hermes</Text>
            </View>
          )}
          <View style={styles.body}>
            <View style={styles.sectionContainer}>
              <Text style={styles.sectionTitle}>Step One</Text>
              <Text style={styles.sectionDescription}>
                Edit <Text style={styles.highlight}>App.js</Text> to change this
                screen and then come back to see your edits.
              </Text>
            </View>
            <View style={styles.sectionContainer}>
              <Text style={styles.sectionTitle}>See Your Changes</Text>
              <Text style={styles.sectionDescription}>
                <ReloadInstructions />
              </Text>
            </View>
            <View style={styles.sectionContainer}>
              <Text style={styles.sectionTitle}>Debug</Text>
              <Text style={styles.sectionDescription}>
                <DebugInstructions />
              </Text>
            </View>
            <View style={styles.sectionContainer}>
              <Text style={styles.sectionTitle}>Learn More</Text>
              <Text style={styles.sectionDescription}>
                Read the docs to discover what to do next:
              </Text>
            </View>
            <LearnMoreLinks />
          </View>
        </ScrollView>
      </SafeAreaView>
    </>
  );
};

const styles = StyleSheet.create({
  scrollView: {
    backgroundColor: Colors.lighter,
  },
  engine: {
    position: 'absolute',
    right: 0,
  },
  body: {
    backgroundColor: Colors.white,
  },
  sectionContainer: {
    marginTop: 32,
    paddingHorizontal: 24,
  },
  sectionTitle: {
    fontSize: 24,
    fontWeight: '600',
    color: Colors.black,
  },
  sectionDescription: {
    marginTop: 8,
    fontSize: 18,
    fontWeight: '400',
    color: Colors.dark,
  },
  highlight: {
    fontWeight: '700',
  },
  footer: {
    color: Colors.dark,
    fontSize: 12,
    fontWeight: '600',
    padding: 4,
    paddingRight: 12,
    textAlign: 'right',
  },
});

export default App;
```

#### index.js
```js
/**
 * @format
 */

import {AppRegistry} from 'react-native';
import App from './App';
import {name as appName} from './app.json';
import messaging from '@react-native-firebase/messaging';
import notifee, { AndroidImportance } from '@notifee/react-native';

async function onMessageReceived(message) {
  await notifee.createChannel({
    id: 'default',
    name: 'Default',
    importance: AndroidImportance.HIGH,
  });

  notifee.displayNotification(JSON.parse(message.data.notifee));

  notifee.incrementBadgeCount();
}

notifee.setBadgeCount(0);

messaging().requestPermission();

if (!messaging().isDeviceRegisteredForRemoteMessages) {
  messaging().registerDeviceForRemoteMessages();
}

messaging().onMessage(onMessageReceived);
messaging().setBackgroundMessageHandler(onMessageReceived);

AppRegistry.registerComponent(appName, () => App);
```
</details>

### iOS

<details><summary>Click To Expand</summary>
<p>

#### `ios/Podfile`:

- [ ] I'm not using Pods
- [x] I'm using Pods and my Podfile looks like:

```ruby
require_relative '../node_modules/react-native/scripts/react_native_pods'
require_relative '../node_modules/@react-native-community/cli-platform-ios/native_modules'

platform :ios, '10.0'

target 'TestPN' do
  config = use_native_modules!

  use_react_native!(:path => config["reactNativePath"])

  target 'TestPNTests' do
    inherit! :complete
    # Pods for testing
  end

  # Enables Flipper.
  #
  # Note that if you have use_frameworks! enabled, Flipper will not work and
  # you should disable these next few lines.
  use_flipper!
  post_install do |installer|
    flipper_post_install(installer)
  end
end

target 'TestPN-tvOS' do
  # Pods for TestPN-tvOS

  target 'TestPN-tvOSTests' do
    inherit! :search_paths
    # Pods for testing
  end
end
```

#### `AppDelegate.m`:

```objc
#import "AppDelegate.h"

#import <React/RCTBridge.h>
#import <React/RCTBundleURLProvider.h>
#import <React/RCTRootView.h>

#ifdef FB_SONARKIT_ENABLED
#import <FlipperKit/FlipperClient.h>
#import <FlipperKitLayoutPlugin/FlipperKitLayoutPlugin.h>
#import <FlipperKitUserDefaultsPlugin/FKUserDefaultsPlugin.h>
#import <FlipperKitNetworkPlugin/FlipperKitNetworkPlugin.h>
#import <SKIOSNetworkPlugin/SKIOSNetworkAdapter.h>
#import <FlipperKitReactPlugin/FlipperKitReactPlugin.h>

#import <Firebase.h>

static void InitializeFlipper(UIApplication *application) {
  FlipperClient *client = [FlipperClient sharedClient];
  SKDescriptorMapper *layoutDescriptorMapper = [[SKDescriptorMapper alloc] initWithDefaults];
  [client addPlugin:[[FlipperKitLayoutPlugin alloc] initWithRootNode:application withDescriptorMapper:layoutDescriptorMapper]];
  [client addPlugin:[[FKUserDefaultsPlugin alloc] initWithSuiteName:nil]];
  [client addPlugin:[FlipperKitReactPlugin new]];
  [client addPlugin:[[FlipperKitNetworkPlugin alloc] initWithNetworkAdapter:[SKIOSNetworkAdapter new]]];
  [client start];
}
#endif

@implementation AppDelegate

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
  if ([FIRApp defaultApp] == nil) {
    [FIRApp configure];
  }
  [FIRMessaging messaging].autoInitEnabled = YES;
  #ifdef FB_SONARKIT_ENABLED
    InitializeFlipper(application);
  #endif

  RCTBridge *bridge = [[RCTBridge alloc] initWithDelegate:self launchOptions:launchOptions];
  RCTRootView *rootView = [[RCTRootView alloc] initWithBridge:bridge
                                                   moduleName:@"TestPN"
                                            initialProperties:nil];

  rootView.backgroundColor = [[UIColor alloc] initWithRed:1.0f green:1.0f blue:1.0f alpha:1];

  self.window = [[UIWindow alloc] initWithFrame:[UIScreen mainScreen].bounds];
  UIViewController *rootViewController = [UIViewController new];
  rootViewController.view = rootView;
  self.window.rootViewController = rootViewController;
  [self.window makeKeyAndVisible];
  return YES;
}

- (NSURL *)sourceURLForBridge:(RCTBridge *)bridge
{
#if DEBUG
  return [[RCTBundleURLProvider sharedSettings] jsBundleURLForBundleRoot:@"index" fallbackResource:nil];
#else
  return [[NSBundle mainBundle] URLForResource:@"main" withExtension:@"jsbundle"];
#endif
}

@end
```

Capabilities eanbled
![image](https://user-images.githubusercontent.com/13423169/87626753-8c3f2800-c703-11ea-92ee-d7c1b1ff3e08.png)

Permissions are enabled
![image](https://user-images.githubusercontent.com/13423169/87627105-60707200-c704-11ea-840c-4be5506414f1.png)

</p>
</details>

---

### Android

<details><summary>Click To Expand</summary>
<p>

#### Have you converted to AndroidX?

- [ ] my application is an AndroidX application?
- [ ] I am using `android/gradle.settings` `jetifier=true` for Android compatibility?
- [ ] I am using the NPM package `jetifier` for react-native compatibility?

#### `android/build.gradle`:

```groovy
// N/A
```

#### `android/app/build.gradle`:

```groovy
// N/A
```

#### `android/settings.gradle`:

```groovy
// N/A
```

#### `MainApplication.java`:

```java
// N/A
```

#### `AndroidManifest.xml`:

```xml
<!-- N/A -->
```

</p>
</details>

---

## Environment

<details><summary>Click To Expand</summary>
<p>

**`react-native info` output:**

```
System:
    OS: macOS 10.15.4
    CPU: (12) x64 Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz
    Memory: 603.45 MB / 16.00 GB
    Shell: 5.7.1 - /bin/zsh
  Binaries:
    Node: 12.16.1 - ~/.nvm/versions/node/v12.16.1/bin/node
    Yarn: 1.9.4 - /usr/local/bin/yarn
    npm: 6.14.4 - ~/.nvm/versions/node/v12.16.1/bin/npm
    Watchman: 4.9.0 - /usr/local/bin/watchman
  Managers:
    CocoaPods: 1.9.3 - /usr/local/bin/pod
  SDKs:
    iOS SDK:
      Platforms: iOS 13.5, DriverKit 19.0, macOS 10.15, tvOS 13.4, watchOS 6.2
    Android SDK:
      API Levels: 23, 25, 26, 27, 28, 29
      Build Tools: 23.0.1, 23.0.3, 25.0.0, 25.0.1, 25.0.2, 25.0.3, 26.0.1, 26.0.2, 26.0.3, 27.0.3, 28.0.2, 28.0.3, 29.0.2
      System Images: android-24 | Google APIs Intel x86 Atom, android-27 | Google APIs Intel x86 Atom
      Android NDK: Not Found
  IDEs:
    Android Studio: 3.5 AI-191.8026.42.35.5791312
    Xcode: 11.5/11E608c - /usr/bin/xcodebuild
  Languages:
    Java: 1.8.0_211 - /usr/bin/javac
    Python: 2.7.15 - /usr/local/bin/python
  npmPackages:
    @react-native-community/cli: Not Found
    react: 16.13.1 => 16.13.1 
    react-native: 0.63.1 => 0.63.1 
  npmGlobalPackages:
    *react-native*: Not Found
```

- **Platform that you're experiencing the issue on**:
  - [x] iOS
  - [ ] Android
  - [x] **iOS** but have not tested behavior on Android
  - [ ] **Android** but have not tested behavior on iOS
  - [ ] Both
- **`react-native-firebase` version you're using that has this issue:**
  - `7.4.2`
- **`Firebase` module(s) you're using that has the issue:**
  - `Messaging`
- **Are you using `TypeScript`?**
  - `N`

</p>
</details>

## Postman
<details><summary>Click To Expand</summary>
<p>

Postman request (Added all the content_available possibilities as I was. not sure which one. should I use)
```
{
	"to" : "dKxdtiyWmUDCqt-0m0GTmF:APA91bFYwxxNrrjOE1MxbCTHcl6hMu7h6eabECZBlSD_DpVufesvsw7BkToRJ_v9rJEmPxklAwOS4sdPujYQGXW92bQ9r8g4t-LrAYm5PbKBjjNLXX6CfT-mA3hRFkKrhcA11UxmMfDh",
	"priority": "high",
	"content_available": true,
	"content-available": true,
	"aps" : {
		"content_available" : 1
	},
	"data": {
	    "notifee": {
	      "title": "Test",
	      "subtitle": "Test",
	      "body": "This message was sent via FCM! 5",
	      "android": {
	        "channelId": "default",
	        "importance": 4,
	        "groupId": "1234"
	      }
	 }
  }
}
```
</p>
</details>

---

- 👉 Check out [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter for updates on the library.

**Comment by mikehardy** [comment]
You should probably remove the device token from your post, though it's not the worst sort of security exposure as they rotate.

Have you done this while plugged into the real device, running the app in release mode from Xcode?

You might not be getting woken up on background because it's power throttled. Working while "killed" (force close, swipe away etc) is not something iOS does for anyone.

**Comment by forbesgillikin** [comment]
Your POST request is incorrect.

Please familiarize yourself with the following docs: 

APS JSON Payload: https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/generating_a_remote_notification

APNS headers: https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/sending_notification_requests_to_apns/#2947607

Most importantly, sending POST requests via the FCM v1 HTTP API:
https://firebase.google.com/docs/cloud-messaging/send-message#send-messages-to-specific-devices

It looks like your request is using the Legacy HTTP method. Please consider migrating to FCM v1 HTTP API as the Firebase documentation mostly supports the v1 API
https://firebase.google.com/docs/cloud-messaging/migrate-v1

Ideally, you would be using the Node Admin SDK to send messages via FCM
https://firebase.google.com/docs/admin/setup


Example POST request using the **Legacy HTTP** method
**A regular alert style notification:**
```javascript
{
    "to": "" // your device fcm token,
     // notice the notification key here, this is what triggers the regular 
     // notification functionality, even when the app is in the background
    "notification": {
        "title": "Notification Title",
        "body": "Notification Body"
    },
   // this data object is available even in the background
    "data": {
       // the data object in FCM only takes string as values
      "notifee": "true",
      "androidChannelID": "default"
    }
}
```

**A silent background notification:**
```javascript
{
    "to": "" // your device fcm token,
    "data": {
        "anyDataKey": "anyValue"
    },
    "apns":{
        "payload": {
            "aps": {
              "contentAvailable": true
            }
          },
          "headers": {
            "apns-push-type": "background",
            "apns-priority": "5",
            "apns-topic": "" // your app bundle identifier
          }
    }
}
```


You must also understand the difference between a **silent** background notification and regular background notification on iOS. Regular background notifications can still contain data but do not need the `content_available` flag. The `content_available` flag is only necessary for silent background notification

**Comment by harrydema** [comment]
@mikehardy I want to send always "silent" notifications and wake up the setBackgroundMessageHandler which is in charge of displaying the notification with notifee (Invertase's notification handler).
You said "Working while "killed" (force close, swipe away etc) is not something iOS does for anyone.", is this true?
I am trying to build a chat app and I need all notifications to be displayed, no matter what state (Foreground, background or quit)

**Comment by mikehardy** [comment]
I state what I think is true. If you research and find otherwise that after force-close something is possible on iOS, I will be interested.

**Comment by harrydema** [comment]
As stated in the documentation, if the app is on quit mode it should call the setBackgroundMessageHandler function anyway:

![image](https://user-images.githubusercontent.com/13423169/87828202-d551c200-c852-11ea-951f-51fb4b41fb11.png)

**Comment by mikehardy** [comment]
Precision is important in terms. Note I am using force-close and also '(force close, swipe away etc) '. I believe my statement is accurate (app is dead dead) for that state on iOS. Quit should still be working, that's when the app is either just not started yet (after boot) or the OS has terminated it (but not the user)

**Comment by forbesgillikin** [comment]
@harrydema @mikehardy is correct here. iOS uses the `application(_:didReceiveRemoteNotification:fetchCompletionHandler:)` instance method to deliver background notifications. In the [iOS docs for application(_:didReceiveRemoteNotification:fetchCompletionHandler:)](https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1623013-application) you can see this statement in the Discussion section

> Use this method to process incoming remote notifications for your app. Unlike the application(_:didReceiveRemoteNotification:) method, which is called only when your app is running in the foreground, the system calls this method when your app is running in the foreground or background. In addition, if you enabled the remote notifications background mode, the system launches your app (or wakes it from the suspended state) and puts it in the background state when a remote notification arrives. **However, the system does not automatically launch your app if the user has force-quit it. In that situation, the user must relaunch your app or restart the device before the system attempts to launch your app automatically again.**

The only way to deliver notifications to apps that have been force-quit by the user is to use the default Notifications system.

**Comment by harrydema** [comment]
I dont completely understand the difference between the force close and the quit state.

If the user swipes up the app i would say this is the quit mode. How is it that the user can force close the app?

**Comment by mikehardy** [comment]
@harrydema you should research more, there are concrete documents on this subject from Apple. "quit" mode is when the user stops using your app, and the operating system kills it to free up resources. Swiping away is force closing. But don't take my word, read the apple docs, it's important to be familiar with them if you develop for iOS as the platform is strict, and subtle.

**Comment by mikehardy** [comment]
As part of the work on #4104 several PRs have been merged, including one just merged and released now as [`@react-native-firebase/7.8.4`](https://github.com/invertase/react-native-firebase/blob/master/packages/messaging/CHANGELOG.md#784-2020-08-31)

We believe FCM messages will be handled now:
- [ if you set `content-available` and `priority` headers correctly](https://github.com/invertase/react-native-firebase/blob/master/packages/messaging/CHANGELOG.md#784-2020-08-31) in your JSON payload 
- if you are on a real device
- if the app is not running and your javascript bundle boots within 8 seconds (careful, redux-saga users! from a patch a few days ago)
- if the app is in the background (with the most recent patch)

We are no longer aware of cases that don't have known solutions.

If you can reproduce a case after updating to the latest code and running `npx react-native-clean-project` to make sure you have a clean build, please open a new issue post an App.js that demonstrates the problem.
